doc/manual/metaprogramming.rst

```diff
@@ -9,106 +9,120 @@
 from within the language, it is possible for a program to transform and
 generate its own code. This allows sophisticated code generation without
-extra build steps, and also allows true Lisp-style macros, as compared
-to preprocessor "macro" systems, like that of C and C++, that perform
-superficial textual manipulation as a separate pass before any real
-parsing or interpretation occurs. Another aspect of metaprogramming is
-reflection: the ability of a running program to dynamically discover
-properties of itself. Reflection emerges naturally from the fact that
-all data types and code are represented by normal Julia data structures,
-so the structure of the program and its types can be explored
-programmatically just like any other data.
-
-Expressions and :func:`eval`
-----------------------------
-
-Julia code is represented as a syntax tree built out of Julia data
-structures of type :obj:`Expr`. This makes it easy to construct and
-manipulate Julia code from within Julia, without generating or parsing
-source text. Here is the definition of the :obj:`Expr` type::
-
-    type Expr
-      head::Symbol
-      args::Array{Any,1}
-      typ
-    end
+extra build steps, and also allows true Lisp-style macros operating at
+the level of abstract syntax. In contrast, preprocessor "macro" systems,
+like that of C and C++, perform textual manipulation and substitution
+before any actual parsing or interpretation occurs.
 
-The ``head`` is a symbol identifying the kind of expression, and
-``args`` is an array of subexpressions, which may be symbols referencing
-the values of variables at evaluation time, may be nested :obj:`Expr`
-objects, or may be actual values of objects. The ``typ`` field is used
-by type inference to store type annotations, and can generally be
-ignored.
+Program representation
+----------------------
 
-There is special syntax for "quoting" code (analogous to quoting
-strings) that makes it easy to create expression objects without
-explicitly constructing :obj:`Expr` objects. There are two forms: a short
-form for inline expressions using ``:`` followed by a single expression,
-and a long form for blocks of code, enclosed in ``quote ... end``. Here
-is an example of the short form used to quote an arithmetic expression:
+Every Julia program starts life as a string:
 
 .. doctest::
 
-    julia> ex = :(a+b*c+1)
-    :(a + b * c + 1)
+    julia> prog = "1 + 1"
+    "1 + 1"
+
+**What happens next?**
+
+The next step is to `parse <http://en.wikipedia.org/wiki/Parsing#Computer_languages>`_
+each string into an object called an expression, represented by the Julia type
+:obj:`Expr`:
+
+.. doctest::
+
+    julia> ex1 = parse(prog)
+    :(1 + 1)
 
     julia> typeof(ex)
     Expr
 
-    julia> ex.head
+Expressions may also be constructed directly in
+`prefix notation <http://en.wikipedia.org/wiki/Polish_notation>`_:
+
+.. doctest::
+
+    julia> ex2 = Expr(:call, :+, 1, 1)
```

Comment:
I don't think `Symbol`s have been introduced earlier in the manual, so somewhere around here might be a good time. Perhaps something as simple as

> `Symbol`s are a special kind of string literal, that can only take values that are valid identifiers *(is that true?)*, denoted by prefixing the string literal with a `:` instead of wrapping it in citation marks. You can also construct symbols from run-time values using `symbol(your_string)`.

I don't know if it's a good idea to interrupt this example, though, so we might want to splice this in somewhere a little earlier.